File: backend/solver/src/timetable_solver/solver_adapter.py
"""Versioned NestJS-to-Python solver adapter envelope."""

import hashlib
import json
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, model_validator

from .contracts import SolveJobRequest

SOLVER_ADAPTER_CONTRACT_VERSION = "SOLVER-ADAPTER-1.0.0"


class SolverAdapterSource(BaseModel):
    model_config = ConfigDict(extra="forbid")

    schemaVersion: Literal["1.0"]
    templateVersion: str = Field(min_length=1)
    schoolId: str = Field(min_length=1)
    academicPeriodId: str = Field(min_length=1)
    ruleSnapshotId: str = Field(min_length=1)
    ruleSetVersion: str = Field(pattern=r"^RULE-SET-[0-9]+\.[0-9]+\.[0-9]+$")
    ruleSnapshotHash: str = Field(pattern=r"^[0-9a-f]{64}$")


class SolverAdapterReproducibility(BaseModel):
    model_config = ConfigDict(extra="forbid")

    randomSeed: int
    timeLimitSeconds: float = Field(gt=0)


class SolverAdapterPayload(BaseModel):
    model_config = ConfigDict(extra="forbid")

    adapterContractVersion: Literal["SOLVER-ADAPTER-1.0.0"]
    source: SolverAdapterSource
    reproducibility: SolverAdapterReproducibility
    input: SolveJobRequest
    inputChecksum: str = Field(pattern=r"^[0-9a-f]{64}$")

    @model_validator(mode="after")
    def validate_checksum(self) -> "SolverAdapterPayload":
        expected = compute_solver_adapter_checksum(self)
        if self.inputChecksum != expected:
            raise ValueError("inputChecksum does not match the canonical adapter payload")
        if self.source.schoolId != self.input.schoolId:
            raise ValueError("adapter source schoolId must match input schoolId")
        if self.source.schemaVersion != self.input.schemaVersion:
            raise ValueError("adapter source schemaVersion must match input schemaVersion")
        return self


def canonicalize(value: object) -> object:
    if isinstance(value, list):
        return [canonicalize(item) for item in value]
    if isinstance(value, dict):
        return {
            key: canonicalize(nested)
            for key, nested in sorted(value.items())
            if nested is not None
        }
    if isinstance(value, float) and value.is_integer():
        return int(value)
    return value


def _unsigned_payload(payload: SolverAdapterPayload) -> dict[str, object]:
    return {
        "adapterContractVersion": payload.adapterContractVersion,
        "source": payload.source.model_dump(mode="json", exclude_none=True),
        "reproducibility": payload.reproducibility.model_dump(mode="json", exclude_none=True),
        "input": payload.input.model_dump(mode="json", exclude_none=True),
    }


def compute_solver_adapter_checksum(payload: SolverAdapterPayload | dict[str, object]) -> str:
    if isinstance(payload, SolverAdapterPayload):
        unsigned = _unsigned_payload(payload)
    else:
        unsigned = {key: value for key, value in payload.items() if key != "inputChecksum"}
    canonical = json.dumps(canonicalize(unsigned), ensure_ascii=False, separators=(",", ":"))
    return hashlib.sha256(canonical.encode("utf-8")).hexdigest()

File: backend/src/contracts/solver-adapter.ts
import { createHash } from "node:crypto";
import type { SolveJobRequest } from "./index";

export const SOLVER_ADAPTER_CONTRACT_VERSION = "SOLVER-ADAPTER-1.0.0" as const;
export const DEFAULT_SOLVER_TIME_LIMIT_SECONDS = 10;

export interface SolverAdapterSource {
  schemaVersion: "1.0";
  templateVersion: string;
  schoolId: string;
  academicPeriodId: string;
  ruleSnapshotId: string;
  ruleSetVersion: string;
  ruleSnapshotHash: string;
}

export interface SolverAdapterReproducibility {
  randomSeed: number;
  timeLimitSeconds: number;
}

export interface SolverAdapterPayload {
  adapterContractVersion: typeof SOLVER_ADAPTER_CONTRACT_VERSION;
  source: SolverAdapterSource;
  reproducibility: SolverAdapterReproducibility;
  input: SolveJobRequest;
  inputChecksum: string;
}

export interface SolverAdapterContext {
  academicPeriodId: string;
  templateVersion: string;
  randomSeed?: number;
  timeLimitSeconds?: number;
}

function canonicalize(value: unknown): unknown {
  if (Array.isArray(value)) return value.map(canonicalize);
  if (value && typeof value === "object") {
    return Object.fromEntries(
      Object.entries(value as Record<string, unknown>)
        .filter(([, nested]) => nested !== null && nested !== undefined)
        .sort(([left], [right]) => (left < right ? -1 : left > right ? 1 : 0))
        .map(([key, nested]) => [key, canonicalize(nested)]),
    );
  }
  return value;
}

function unsignedPayload(payload: Omit<SolverAdapterPayload, "inputChecksum">) {
  return canonicalize(payload);
}

export function computeSolverAdapterChecksum(payload: Omit<SolverAdapterPayload, "inputChecksum">) {
  return createHash("sha256")
    .update(JSON.stringify(unsignedPayload(payload)))
    .digest("hex");
}

export function buildSolverAdapterPayload(input: SolveJobRequest, context: SolverAdapterContext): SolverAdapterPayload {
  const ruleSnapshotId = input.ruleSnapshotId;
  const ruleSetVersion = input.ruleSetVersion;
  const ruleSnapshotHash = input.ruleSnapshotHash;
  if (!ruleSnapshotId || !ruleSetVersion || !ruleSnapshotHash) {
    throw new Error("Bộ điều hợp tối ưu yêu cầu tham chiếu bản chụp quy tắc đã phê duyệt.");
  }
  if (!context.academicPeriodId || !context.templateVersion) {
    throw new Error("Bộ điều hợp tối ưu yêu cầu academicPeriodId và templateVersion.");
  }

  const unsigned = {
    adapterContractVersion: SOLVER_ADAPTER_CONTRACT_VERSION,
    source: {
      schemaVersion: input.schemaVersion,
      templateVersion: context.templateVersion,
      schoolId: input.schoolId,
      academicPeriodId: context.academicPeriodId,
      ruleSnapshotId,
      ruleSetVersion,
      ruleSnapshotHash,
    },
    reproducibility: {
      randomSeed: context.randomSeed ?? 0,
      timeLimitSeconds:
        context.timeLimitSeconds ?? input.options?.timeLimitSeconds ?? DEFAULT_SOLVER_TIME_LIMIT_SECONDS,
    },
    input,
  } satisfies Omit<SolverAdapterPayload, "inputChecksum">;

  return {
    ...unsigned,
    inputChecksum: computeSolverAdapterChecksum(unsigned),
  };
}

export function verifySolverAdapterChecksum(payload: SolverAdapterPayload) {
  const { inputChecksum: received, ...unsigned } = payload;
  return received === computeSolverAdapterChecksum(unsigned);
}

export function validateSolverAdapterPayload(payload: SolverAdapterPayload) {
  if (payload.adapterContractVersion !== SOLVER_ADAPTER_CONTRACT_VERSION) {
    throw new Error(`Hợp đồng bộ điều hợp tối ưu không được hỗ trợ: ${payload.adapterContractVersion}`);
  }
  if (!/^[0-9a-f]{64}$/.test(payload.inputChecksum) || !verifySolverAdapterChecksum(payload)) {
    throw new Error("inputChecksum của bộ điều hợp tối ưu không khớp với dữ liệu chuẩn.");
  }
  return payload;
}
